sys/vm/swap_pager.c

```diff
@@ -2207,18 +2208,29 @@ swp_pager_meta_cheri_get_tags(vm_page_t page)
 	    i < (swidx + 1) * BITS_PER_TAGS_PER_PAGE; i++) {
 		p = scan;
 		for (t = sb->swb_tags[i]; t != 0; t >>= j) {
+			mark_capdirty = 1;
 			j = ffsl((long)t);
 			cheri_restore_tag(p + j - 1);
 			p += j;
 		}
 		scan += 8 * sizeof(uint64_t);
 	}
+
+	/*
+	 * Because we stored through the direct region we may have bypassed
+	 * the MMU and all its implicit capdirty tracking.
```

Comment:
Why isn't this a problem for normal dirty?

Reply:
Having just restored the page from swap, it is *clean* in the normal sense of dirty: it matches its backing store and can be evicted without needing to first write back.

Reply:
While true, that's probably fine, I think?  In fact, we'd probably like to notice that a paged-out page is logically `!PGA_CAPSTORE` before we haul it back in for revocation, but that's a different chunk of work.

Reply:
The intention of `capstore_on_alloc` was to measure the benefit of assuming that `HASCAP` pages would quickly come to have capabilities even if their allocation wasn't due to a `VM_PROT_WRITE_CAP` fault; roughly it'd be like assuming that `VM_PROT_WRITE` pages are likely to be dirtied soon after their allocation.  I conjecture that it's at least a little important that `PGA_CAPSTORE` can be clear for `HASCAP` objects' pages, tho', so that the revoker can skip pages that might, but don't (and haven't for a while) had capabilities on them.

Reply:
Do you clear `PGA_CAPSTORE` after sweeping a page?  What I find odd is that swapping a page can have different semantics than if the page just happens to not get swapped out.  Given how rare swapping is in much of our CHERI test environments, I worry that such semantic differences will not be tested very often if they only occur as a result of swap.

Reply:
There's a rather detailed dance involving `PGA_CAPSTORE`, `PGA_CAPDIRTY`, and the MD PTE bits so that we don't need to do IPI TLB invalidations during a revocation pass and yet can eventually stop scanning cap-free pages.  But yes, the revoker can clear `CAPSTORE` after a pass if it `xbusy`'d the page, found no capabilities while visiting the page, there are no PTEs with `CW` asserted (`CD` asserted implies `CW` asserted), and `CAPDIRTY` is clear.  Because the page has been `xbusy`, no new mappings can have been created, but mappings may have been torn down, as the pmap is not locked during the visit, moving the set PTE `CD` to a set `PGA_CAPDIRTY` flag.

I should probably point you at the report in CTSRD svn `reports/2020-revocation-arch-tech`, with apologies for the optimistic-in-retrospect name.

Reply:
Ok, if we can get the same semantics without swap I'm fine with having this be lazy for swap.